### PR TITLE
Use memcpy/memcmp for unaligned reads.

### PR DESCRIPTION
**Pull request by nmoinvaz**
Redid #934 using macros suggested in that PR.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
I can't seem to figure out why PPC doesn't work. Seems to be decompression which might suggest chunkmemset code.

**Comment by nmoinvaz** [comment]
### DEVELOP https://github.com/zlib-ng/zlib-ng/commit/63bc287e4db74a62940876143c3c6f9058f401de
```
 Tool: ../zlib-ng/build-develop/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.809%      1.047/1.059/1.065/0.006        0.555/0.564/0.570/0.003      104,067,822
 2     35.391%      1.745/1.760/1.770/0.007        0.569/0.578/0.583/0.004       82,195,416
 3     33.698%      2.436/2.447/2.457/0.006        0.548/0.557/0.564/0.005       78,262,436
 4     32.989%      2.709/2.739/2.752/0.010        0.531/0.541/0.547/0.005       76,616,183
 5     32.483%      2.820/2.837/2.846/0.008        0.527/0.537/0.542/0.004       75,441,984
 6     32.318%      3.435/3.461/3.474/0.011        0.518/0.530/0.535/0.004       75,058,801
 7     32.061%      4.798/4.842/4.860/0.015        0.524/0.533/0.538/0.004       74,461,300
 8     31.978%      7.382/7.417/7.439/0.016        0.520/0.530/0.536/0.004       74,269,032
 9     31.727%      9.520/9.591/9.621/0.030        0.518/0.527/0.533/0.004       73,685,100

 avg1  34.161%                        4.017                          0.544
 tot                               1084.580                        146.896      714,058,074

 Tool: ../zlib-ng/build-develop/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.809%      1.043/1.055/1.063/0.006        0.559/0.565/0.572/0.004      104,067,822
 2     35.391%      1.743/1.754/1.763/0.006        0.568/0.578/0.586/0.006       82,195,416
 3     33.698%      2.406/2.434/2.447/0.010        0.547/0.557/0.564/0.006       78,262,436
 4     32.989%      2.717/2.733/2.746/0.008        0.527/0.540/0.548/0.005       76,616,183
 5     32.483%      2.813/2.830/2.840/0.007        0.524/0.534/0.541/0.005       75,441,984
 6     32.318%      3.428/3.447/3.463/0.010        0.521/0.528/0.536/0.005       75,058,801
 7     32.061%      4.792/4.830/4.850/0.015        0.522/0.532/0.539/0.004       74,461,300
 8     31.978%      7.368/7.398/7.418/0.014        0.519/0.530/0.537/0.004       74,269,032
 9     31.727%      9.507/9.550/9.580/0.023        0.518/0.528/0.533/0.004       73,685,100

 avg1  34.161%                        4.004                          0.544
 tot                               1080.978                        146.752      714,058,074
```
### PR https://github.com/zlib-ng/zlib-ng/pull/1086/commits/ace9a29c590b3aeed95bf1b4eb153ecfd4b06fd8
```
 Tool: ../zlib-ng/build-memalign/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.809%      1.060/1.075/1.086/0.008        0.531/0.541/0.549/0.006      104,067,822
 2     35.391%      1.733/1.751/1.765/0.009        0.547/0.557/0.567/0.006       82,195,416
 3     33.698%      2.418/2.444/2.468/0.013        0.529/0.540/0.547/0.005       78,262,436
 4     32.989%      2.720/2.746/2.771/0.012        0.511/0.523/0.531/0.005       76,616,183
 5     32.483%      2.877/2.906/2.937/0.015        0.505/0.519/0.526/0.006       75,441,984
 6     32.318%      3.491/3.524/3.556/0.016        0.502/0.514/0.522/0.004       75,058,801
 7     32.061%      4.803/4.840/4.873/0.021        0.504/0.517/0.526/0.006       74,461,300
 8     31.978%      7.365/7.405/7.435/0.017        0.504/0.514/0.522/0.006       74,269,032
 9     31.727%      9.501/9.565/9.614/0.033        0.499/0.508/0.514/0.005       73,685,100

 avg1  34.161%                        4.028                          0.526
 tot                               1087.652                        142.024      714,058,074

 Tool: ../zlib-ng/build-memalign/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.809%      1.056/1.068/1.075/0.005        0.530/0.538/0.543/0.004      104,067,822
 2     35.391%      1.726/1.741/1.750/0.008        0.542/0.554/0.561/0.005       82,195,416
 3     33.698%      2.413/2.437/2.448/0.010        0.527/0.533/0.537/0.003       78,262,436
 4     32.989%      2.706/2.732/2.743/0.010        0.510/0.516/0.521/0.003       76,616,183
 5     32.483%      2.878/2.897/2.907/0.007        0.504/0.514/0.519/0.004       75,441,984
 6     32.318%      3.492/3.514/3.525/0.009        0.499/0.507/0.514/0.005       75,058,801
 7     32.061%      4.800/4.828/4.840/0.011        0.502/0.512/0.518/0.004       74,461,300
 8     31.978%      7.366/7.395/7.411/0.012        0.505/0.509/0.513/0.002       74,269,032
 9     31.727%      9.499/9.558/9.583/0.022        0.496/0.502/0.507/0.003       73,685,100

 avg1  34.161%                        4.019                          0.521
 tot                               1085.102                        140.564      714,058,074
```

Compiler is MSVC 19.30.30706.0.

**Comment by nmoinvaz** [comment]
It appears that MSVC requires either `/O2` or `/Ox` for it to convert `memcmp` or `memcpy` to simple instructions. I need to check if those are enabled in release builds. For GCC and Clang, it doesn't require optimization flags to make this trade.

[Here is a compiler explorer](https://godbolt.org/z/E3v99Gdss) that shows each compiler and what it does.

**Comment by nmoinvaz** [comment]
### DEVELOP https://github.com/zlib-ng/zlib-ng/commit/63bc287e4db74a62940876143c3c6f9058f401de
```
 Tool: ../zlib-ng/build-develop/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     96.147%      1.738/1.832/1.856/0.027        0.702/0.715/0.721/0.005      223,301,805
 2     58.252%      3.560/3.580/3.596/0.010        0.840/0.917/0.945/0.034      135,290,814
 3     58.199%      4.427/4.486/4.521/0.025        0.853/0.896/0.921/0.020      135,166,436
 4     55.318%      9.303/9.342/9.372/0.021        0.841/0.863/0.872/0.007      128,475,121
 5     54.501%   11.701/11.786/11.826/0.033        0.858/0.871/0.877/0.005      126,577,230
 6     48.008%   31.668/31.773/31.888/0.070        0.799/0.807/0.814/0.005      111,497,930
 7     43.962%   54.373/54.571/54.790/0.115        0.817/0.833/0.842/0.007      102,101,531
 8     39.601% 177.614/178.150/178.537/0.261        0.748/0.759/0.765/0.005       91,973,526
 9     31.727%      9.346/9.395/9.429/0.024        0.515/0.521/0.525/0.002       73,685,100

 avg1  53.968%                       33.879                          0.798
 tot                               9147.443                        215.437    1,128,069,493
```
### PR https://github.com/zlib-ng/zlib-ng/pull/1086/commits/ace9a29c590b3aeed95bf1b4eb153ecfd4b06fd8
```

 Tool: ../zlib-ng/build-memalign/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     96.147%      1.729/1.770/1.851/0.041        0.619/0.630/0.636/0.005      223,301,805
 2     58.252%      3.622/3.656/3.686/0.017        0.754/0.773/0.841/0.017      135,290,814
 3     58.199%      4.871/4.943/5.049/0.059        0.761/0.773/0.780/0.006      135,166,436
 4     55.318%      9.088/9.179/9.349/0.087        0.785/0.792/0.800/0.004      128,475,121
 5     54.501%   11.480/11.587/11.753/0.092        0.793/0.807/0.815/0.006      126,577,230
 6     48.008%   31.125/31.422/31.768/0.216        0.751/0.763/0.770/0.005      111,497,930
 7     43.962%   53.668/54.115/54.552/0.298        0.797/0.811/0.817/0.004      102,101,531
 8     39.601% 174.881/176.316/177.997/1.030        0.739/0.747/0.751/0.003       91,973,526
 9     31.727%      9.020/9.124/9.292/0.085        0.506/0.512/0.517/0.003       73,685,100

 avg1  53.968%                       33.568                          0.734
 tot                               9063.359                        198.257    1,128,069,493
```

Compiler is Clang 12.0.0 with MSVC-like command-line.

**Comment by nmoinvaz** [comment]
It looks like for MSVC, cmake defaults to `/O2` in Release mode so I'm not sure why it is producing slower code for compression.

**Comment by nmoinvaz** [comment]
One interesting thing is that ClangCl produces poor results. It looks like this is entirely due to the wrong `sse4` flag being used in CMake. Because of that it does not properly detect the availability of `_mm_crc32_u32` resulting in the code falling-back to the `_asm crc32` which has a bug and I have submitted a PR for.

The issue with the wrong flag being used is fixed in my ClangCl PR https://github.com/zlib-ng/zlib-ng/pull/1083/commits/0372286e750750d2e7dd2e70693b5f448286db98.

**Comment by nmoinvaz** [comment]
Here is updated ClangCl benchmarks after my fixes:

### BASE https://github.com/nmoinvaz/zlib-ng/commit/4a29135dcf6c20b6d661e1d3b2d3ac8c1cf36349
```
 Tool: ../zlib-ng/build-crc32-asm/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.106/0.108/0.109/0.001        0.048/0.049/0.050/0.000        8,523,732
 2     43.871%      0.169/0.170/0.172/0.001        0.050/0.051/0.052/0.001        6,903,609
 3     42.387%      0.198/0.201/0.203/0.001        0.048/0.050/0.051/0.001        6,670,099
 4     41.647%      0.224/0.227/0.229/0.001        0.047/0.048/0.049/0.001        6,553,723
 5     41.216%      0.245/0.247/0.249/0.001        0.046/0.048/0.049/0.001        6,485,938
 6     41.037%      0.282/0.285/0.288/0.002        0.046/0.048/0.049/0.001        6,457,776
 7     40.778%      0.362/0.368/0.372/0.003        0.046/0.048/0.049/0.001        6,416,919
 8     40.704%      0.454/0.459/0.464/0.003        0.046/0.048/0.049/0.001        6,405,244
 9     40.409%      0.490/0.500/0.504/0.004        0.045/0.047/0.048/0.001        6,358,951

 avg1  42.913%                        0.285                          0.049
 tot                                 76.949                         13.105       60,775,991
```
### PR https://github.com/nmoinvaz/zlib-ng/commit/f7dc6e0a7067816414844a2cfc2b50c822249bce
```
 Tool: ../zlib-ng/build-memalign/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.093/0.094/0.095/0.001        0.048/0.050/0.051/0.001        8,523,732
 2     43.871%      0.155/0.157/0.159/0.001        0.050/0.051/0.052/0.000        6,903,609
 3     42.387%      0.183/0.186/0.187/0.001        0.048/0.050/0.051/0.001        6,670,099
 4     41.647%      0.206/0.211/0.213/0.001        0.048/0.049/0.050/0.001        6,553,723
 5     41.216%      0.227/0.231/0.233/0.002        0.046/0.048/0.049/0.001        6,485,938
 6     41.037%      0.265/0.269/0.272/0.002        0.046/0.048/0.049/0.001        6,457,776
 7     40.778%      0.341/0.349/0.353/0.003        0.047/0.048/0.050/0.001        6,416,919
 8     40.704%      0.439/0.445/0.449/0.003        0.047/0.048/0.049/0.001        6,405,244
 9     40.409%      0.493/0.499/0.504/0.004        0.045/0.047/0.049/0.001        6,358,951

 avg1  42.913%                        0.271                          0.049
 tot                                 73.211                         13.184       60,775,991
```
If my calculations `(1.0-73.211/76.949)*100` are correct it is 4.8% improvement in compression speed at least for this small dataset. My previous test only showed 0.9% improvement which is more like what I'd expect.

**Comment by nmoinvaz** [comment]
Updated MSVC benchmarks:

### BASE https://github.com/nmoinvaz/zlib-ng/commit/4a29135dcf6c20b6d661e1d3b2d3ac8c1cf36349
```
 Tool: ../zlib-ng/build-develop/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.095/0.096/0.096/0.001        0.048/0.049/0.049/0.000        8,523,732
 2     43.871%      0.154/0.155/0.156/0.001        0.050/0.050/0.051/0.000        6,903,609
 3     42.387%      0.209/0.211/0.213/0.001        0.048/0.049/0.050/0.000        6,670,099
 4     41.647%      0.231/0.233/0.234/0.001        0.047/0.048/0.049/0.001        6,553,723
 5     41.216%      0.237/0.239/0.241/0.001        0.046/0.048/0.049/0.001        6,485,938
 6     41.037%      0.274/0.278/0.280/0.002        0.046/0.047/0.048/0.001        6,457,776
 7     40.778%      0.347/0.351/0.355/0.002        0.046/0.047/0.048/0.001        6,416,919
 8     40.704%      0.437/0.443/0.447/0.002        0.046/0.047/0.048/0.001        6,405,244
 9     40.409%      0.514/0.519/0.523/0.003        0.045/0.047/0.048/0.001        6,358,951

 avg1  42.913%                        0.281                          0.048
 tot                                 75.744                         12.964       60,775,991
```
### PR https://github.com/nmoinvaz/zlib-ng/commit/f7dc6e0a7067816414844a2cfc2b50c822249bce
```
 Tool: ../zlib-ng/build-memalign/Release/minigzip.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.166%      0.096/0.097/0.098/0.000        0.046/0.047/0.048/0.001        8,523,732
 2     43.871%      0.152/0.154/0.155/0.001        0.048/0.049/0.050/0.001        6,903,609
 3     42.387%      0.209/0.211/0.212/0.001        0.046/0.047/0.048/0.000        6,670,099
 4     41.647%      0.231/0.233/0.234/0.001        0.045/0.046/0.047/0.001        6,553,723
 5     41.216%      0.243/0.245/0.246/0.001        0.045/0.047/0.047/0.001        6,485,938
 6     41.037%      0.277/0.282/0.285/0.002        0.045/0.046/0.046/0.001        6,457,776
 7     40.778%      0.345/0.350/0.354/0.002        0.045/0.046/0.047/0.001        6,416,919
 8     40.704%      0.438/0.443/0.447/0.002        0.045/0.046/0.046/0.000        6,405,244
 9     40.409%      0.512/0.519/0.526/0.004        0.044/0.045/0.047/0.001        6,358,951

 avg1  42.913%                        0.282                          0.047
 tot                                 76.020                         12.589       60,775,991
```
~0.36% decrease in compression speed.

File: chunkset.c [hunk 1]
```diff
@@ -63,13 +60,11 @@ static inline void loadchunk(uint8_t const *s, chunk_t *chunk) {
...
-    *(uint64_t *)out = *chunk;
-#elif defined(UNALIGNED_OK)
-    ((uint32_t *)out)[0] = chunk->u32[0];
-    ((uint32_t *)out)[1] = chunk->u32[1];
+#if defined(UNALIGNED_OK) && !defined(UNALIGNED64_OK)
+    zmemcpy_4(out, &chunk->u32[0]);
+    zmemcpy_4(out+4, &chunk->u32[1]);
 #else
-    memcpy(out, chunk, sizeof(chunk_t));
+    zmemcpy_8(out, chunk);
```
Comment (Dead2) [inline chunkset.c]: Couldn't this use zmemcpy_8 unconditionally?
I might be missing something obvious though, but it looks to me like the fallback `zmemcpy_8` for `UNALIGNED_OK` without `UNALIGNED64_OK` should be doing the same thing as the first condition in this `#if`.
Just not sure whether this is intentional or something forgotten after making the fallback `zmemcpy_8`.

If intentional, would it warrant a comment about it not being compatible?

Comment (nmoinvaz) [reply chunkset.c]: I think you are correct.

Comment (nmoinvaz) [reply chunkset.c]: Fixed.

**Comment by KungFuJesus** [comment]
As mentioned in here: https://github.com/zlib-ng/zlib-ng/issues/1100, memcmp kind of sucks for little endian architectures.  You'd think the builtin that GCC is emitting for it would notice we only ever compare to not 0 and omit the byte swaps that happen.  Maybe it does, compiler explorer is only emitting the result of a generic return rather than a direct comparison with 0.

**Comment by KungFuJesus** [comment]
Huzzah it does:
https://godbolt.org/z/fPrWsrh5W

Disregard my comment.

**Comment by nmoinvaz** [comment]
@KungFuJesus I used your compiler explorer to determine than it was GCC < 7 instead of GCC < 8 where it doesn't optimize memcmp.

**Comment by nmoinvaz** [comment]
I have also added a comment about why we opt for not using memcpy/memcmp for older GCC.

File: zutil.h [hunk 2]
```diff
@@ -261,6 +261,31 @@ void Z_INTERNAL   zng_cfree(void *opaque, void *ptr);
 #  endif
 #endif
 
+/* GCC < 7 doesn't optimize memcmp with unaligned instructions if the architecture
+ * supports it. This can result in negative performance impact of around 36%. */
+#if defined(UNALIGNED_OK) && (defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 7)
```
Comment (nmoinvaz) [inline zutil.h:266]: @Dead2:

- Should we only do this for memcmp? Or for both memcmp and memcpy like it is now? It really only affects memcmp.
- What do you think if we had a `FORCE_UNALIGNED` build option that would enable this in CMake/Configure? Is that preferrable?

Comment (Dead2) [reply zutil.h:266]: If we can avoid doing it for memcpy without a significant performance impact, then there is really nothing to debate.

`FORCE_UNALIGNED` might be a good idea, or `ALLOW_UNALIGNED` perhaps, but I think that change should be in a separate PR in any case, so this one stays a simple cleanup PR and not something requiring a lot of discussion.

**Comment by nmoinvaz** [comment]
Modified the macros to only apply to memcmp and not memcpy since GCC has always converted memcpy to proper instructions according to compiler explorer.

**Comment by Dead2** [comment]
Perhaps we could split up this PR into one for memcmp and one for memcpy. I am seeing some warnings with the memcmp.
For example:
```
/usr/src/zlib-ng/match_tpl.h:151:17: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                 if (zmemcmp_2(mbase_end+cur_match, &scan_end) == 0 &&
                 ^
```

If we could convert the relevant type-punned assignments to plain memcpy in a separate PR, then it should also be easy to verify its correctness and get that applied first. What do you think @nmoinvaz?

**Comment by Dead2** [comment]
Actually there are only 4 such warnings, (repeated for each instance of the template):
Here are all of them:
```
/usr/src/zlib-ng/match_tpl.h:151:17: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                 if (zmemcmp_2(mbase_end+cur_match, &scan_end) == 0 &&
                 ^
/usr/src/zlib-ng/match_tpl.h:152:21: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                     zmemcmp_2(mbase_start+cur_match, &scan_start) == 0)
                     ^
/usr/src/zlib-ng/match_tpl.h:167:17: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                 if (zmemcmp_4(mbase_end+cur_match, &scan_end) == 0 &&
                 ^
/usr/src/zlib-ng/match_tpl.h:168:21: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                     zmemcmp_4(mbase_start+cur_match, &scan_start) == 0)
                     ^
```

**Comment by nmoinvaz** [comment]
I would prefer not to do another PR. Maybe it is because I am casting to (uint8_t *) in zmemcmp_*.

**Comment by Dead2** [comment]
I forgot to say that this is on x86_64 (Xeon X5660) with GCC 4.8. So it should be using the UNALIGNED_OK route

**Comment by mtl1979** [comment]
I did mention it's bad to use casts in preprocessor macros... inline functions are better when casts are required... Due to how modern compilers work, casts need temporary variable to work correctly.

**Comment by nmoinvaz** [comment]
Those warnings should be fixed now. I just have to carry around extra `scan_start` and `scan_end` variables of each type.

**Comment by nmoinvaz** [comment]
@Dead2 would you prefer `zmemcpy` just replaced with `memcpy`?

**Comment by Dead2** [comment]
@nmoinvaz That was what I was thinking, yes. No need to use a wrapper macro if it is always just a standard memcpy after all, right?

I did some "quick" benchmarking of this on the X5660 running GCC4.8:

## Develop
```
 Tool: minigzip   Levels: 1-9
 Runs: 12         Trim worst: 6

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.673/1.703/1.732/0.021        0.828/0.846/0.864/0.012       94,127,290
 2     35.518%      3.058/3.074/3.089/0.013        0.843/0.861/0.889/0.018       75,282,961
 3     33.882%      3.642/3.683/3.696/0.021        0.830/0.844/0.850/0.007       71,816,478
 4     33.174%      4.097/4.115/4.153/0.021        0.808/0.823/0.832/0.011       70,315,668
 5     32.660%      4.587/4.638/4.666/0.030        0.803/0.811/0.820/0.007       69,225,542
 6     32.508%      5.241/5.270/5.305/0.023        0.785/0.792/0.801/0.008       68,902,222
 7     32.255%      6.921/6.958/6.984/0.025        0.801/0.806/0.812/0.004       68,366,800
 8     32.167%   10.505/10.572/10.609/0.037        0.773/0.785/0.796/0.009       68,180,776
 9     31.887%   11.507/11.537/11.563/0.022        0.792/0.811/0.824/0.013       67,586,442

 avg1  34.273%                        5.728                          0.820
 tot                                309.301                         44.280      653,804,179

   text    data     bss     dec     hex filename
 130407    1456      32  131895   20337 libz-ng.so.2
```

## zmemcmp + memcpy simplification (First edition, lost the commit number)
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.695/1.708/1.724/0.011        0.807/0.829/0.843/0.014       94,127,290
 2     35.518%      3.049/3.090/3.113/0.022        0.830/0.854/0.876/0.019       75,282,961
 3     33.882%      3.629/3.675/3.705/0.028        0.827/0.837/0.847/0.009       71,816,478
 4     33.174%      4.080/4.121/4.168/0.041        0.812/0.820/0.826/0.005       70,315,668
 5     32.660%      4.610/4.655/4.681/0.025        0.772/0.797/0.813/0.015       69,225,542
 6     32.508%      5.241/5.293/5.353/0.050        0.768/0.793/0.808/0.016       68,902,222
 7     32.255%      6.981/7.008/7.042/0.020        0.786/0.794/0.805/0.007       68,366,800
 8     32.167%   10.506/10.547/10.562/0.022        0.741/0.778/0.796/0.020       68,180,776
 9     31.887%   11.679/11.721/11.749/0.027        0.760/0.801/0.811/0.020       67,586,442

 avg1  34.273%                        5.758                          0.811
 tot                                310.905                         43.819      653,804,179
 
   text    data     bss     dec     hex filename
 130327    1456      32  131815   202e7 libz-ng.so.2
 130327    1456      32  131815   202e7 libz-ng.so.2.1.0.devel
```

## zmemcmp + zmemcpy macros daf77d64d5588d92bf1bce20d3cb933e4e16ef1c
```
 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     44.409%      1.687/1.704/1.729/0.014        0.849/0.856/0.867/0.007       94,127,290
 2     35.518%      3.055/3.068/3.081/0.012        0.853/0.872/0.880/0.009       75,282,961
 3     33.882%      3.603/3.631/3.654/0.019        0.817/0.835/0.843/0.010       71,816,478
 4     33.174%      4.058/4.096/4.114/0.022        0.789/0.796/0.808/0.007       70,315,668
 5     32.660%      4.566/4.626/4.658/0.032        0.775/0.791/0.803/0.012       69,225,542
 6     32.508%      5.209/5.283/5.333/0.052        0.781/0.806/0.824/0.015       68,902,222
 7     32.255%      6.942/6.996/7.040/0.034        0.762/0.785/0.800/0.014       68,366,800
 8     32.167%   10.475/10.509/10.539/0.022        0.768/0.776/0.787/0.008       68,180,776
 9     31.887%   11.674/11.713/11.767/0.032        0.750/0.794/0.813/0.022       67,586,442

 avg1  34.273%                        5.736                          0.812
 tot                                309.752                         43.865      653,804,179

   text    data     bss     dec     hex filename
 130535    1456      32  132023   203b7 libz-ng.so.2
```

Looks like the compiled size varies quite a bit.
Performance is not too different.

I will have to re-run benchmarks with the newer commits.

**Comment by Dead2** [comment]
## Develop 3adbc36c8aa1ddaeab1f8d2fd7122106d3221889
 ```
 Tool: minigzip   Levels: 1-9
 Runs: 50         Trim worst: 28

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.678/2.713/2.738/0.018        1.178/1.245/1.264/0.020      153,422,996
 2     43.877%      2.435/2.472/2.490/0.013        0.604/0.631/0.648/0.013       62,140,932
 3     42.381%      2.160/2.222/2.241/0.021        0.470/0.482/0.490/0.006       46,684,570
 4     41.665%      2.118/2.135/2.148/0.010        0.378/0.394/0.407/0.008       39,338,830
 5     41.211%      2.362/2.417/2.440/0.021        0.372/0.394/0.403/0.008       38,910,121
 6     41.036%      2.251/2.288/2.305/0.012        0.310/0.322/0.330/0.006       32,287,513
 7     40.784%      2.286/2.317/2.340/0.015        0.229/0.254/0.265/0.011       25,671,718
 8     40.707%      2.108/2.137/2.150/0.010        0.177/0.191/0.197/0.006       19,217,360
 9     40.419%      2.555/2.590/2.603/0.010        0.187/0.195/0.200/0.004       19,081,493

 avg1  42.916%                        2.366                          0.457
 tot                                468.433                         90.423      436,755,533
 
   text    data     bss     dec     hex filename
 130407    1456      32  131895   20337 libz-ng.so.2
 ```
 
## PR #1086 77a44b7cc940ba5874a0bafeb946992751d5dd76
 ```
  Tool: minigzip   Levels: 1-9
 Runs: 50         Trim worst: 28

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     54.165%      2.644/2.716/2.745/0.027        1.203/1.237/1.254/0.014      153,422,996
 2     64.444%      4.797/4.868/4.900/0.026        0.674/0.712/0.727/0.015       91,269,358
 3     64.337%      4.265/4.314/4.339/0.018        0.525/0.551/0.560/0.008       70,870,345
 4     64.004%      5.311/5.379/5.406/0.021        0.457/0.470/0.482/0.007       60,430,944
 5     63.924%      6.198/6.243/6.273/0.026        0.447/0.464/0.472/0.007       60,356,181
 6     63.380%      8.432/8.468/8.493/0.016        0.371/0.386/0.396/0.008       49,868,047
 7     63.280%      8.520/8.551/8.575/0.017        0.285/0.303/0.317/0.009       39,831,828
 8     63.136%      9.504/9.526/9.542/0.010        0.214/0.228/0.236/0.007       29,805,774
 9     63.149%   21.290/21.312/21.323/0.010        0.217/0.228/0.236/0.006       29,811,921

 avg1  62.646%                        7.931                          0.509
 tot                               1570.281                        100.751      585,667,394
 
   text    data     bss     dec     hex filename
 131159    1456      32  132647   20627 libz-ng.so.2
 ```
 
 Whatever happened between c0c2e744da96580b3075240c671925784229221b and 6a4fa870ed1132b99a1f9d35d4a1350d9413a6af?
 Compression ratio is completely wrong for everything except level 1 now, and level9 got super slow.
 
 Handling of scan_endX/scan_startX might be missing something.

**Comment by nmoinvaz** [comment]
I split this PR into a separate `memcpy` PR so those changes can get in there quicker.

**Pull request closed without merging**